subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/impl/DirectorySnapshotterStatisticsTest.groovy

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.snapshot.impl
+
+import org.gradle.api.internal.cache.StringInterner
+import org.gradle.internal.hash.TestFileHasher
+import org.gradle.test.fixtures.file.CleanupTestDirectory
+import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
+import org.gradle.util.Requires
+import org.gradle.util.TestPrecondition
+import org.gradle.util.UsesNativeServices
+import org.junit.Rule
+import spock.lang.Specification
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+@UsesNativeServices
+@CleanupTestDirectory(fieldName = "tmpDir")
+class DirectorySnapshotterStatisticsTest extends Specification {
```

Comment:
Can you reduce the number of test cases in this file? Having one test for
- nested hierarchy
- empty directory
- failed visit
- visit other tree via symlink

should be enough, right?